### ISSUE TITLE
A question about the number of patches (per bag) used to train the model

### ISSUE THREAD

**Issue by wuxchcandoit** [binli123/dsmil-wsi#77]
Hi!
In the code, it seems that all patches of a bag are used to train the model. However, there are in average about 5,000 patches per bag and each patch is represented by an embedding whose dimension is 512. What is the configuration of the graphics card used during training? Has the OOM (Out Of Memory) problem ever arisen?

**Comment by binli123** [comment]
Hi, the test set should be reserved before the training, and the training process further splits the training data into a training set and a validation set. After the training is done you can then evaluate with the best model on the test data.

**Comment by binli123** [comment]
I incorporated the training/testing into the same pipeline in the latest commit. You can set --eval_scheme=5-fold-cv-standalone-test which will perform a train/valid/test like this: 
>A standalone test set consisting of 20% samples is reserved, remaining 80% of samples are used to construct a 5-fold cross-validation.
>For each fold, the best model and corresponding threshold are saved.
>After the 5-fold cross-validation, 5 best models along with the corresponding optimal thresholds are obtained which are used to perform inference on the reserved test set. A final prediction for a test sample is the majority vote of the 5 models.
>For a binary classification, accuracy and balanced accuracy scores are computed. For a multi-label classification, hamming loss (smaller the better) and subset accuracy are computed.

You can also simply run a 5-fold cv --eval_scheme=5-fold-cv

There were some issues with the testing script when loading pretrained weights (i.e., sometimes the weights are not fully loaded or there are missing weights, setting strict=False can reveal the problems.). The purpose of the testing script is to generate the heatmap, you should now read the performance directly from the training script. I will fix the issues in a couple of days.
